openstates/me/people.py

```diff
@@ -31,75 +31,99 @@ def scrape(self, chamber=None):
+        mapping = {
+            'district': 0,
+            'last_name': 1,
+            'first_name': 2,
+            'mid_name': 3,
+            'jr_sr': 4,
+            'party': 5,
+            'legal_res': 6,
+            'address1': 7,
+            'address2': 8,
+            'town': 9,
+            'state': 10,
+            'zip_code': 11,
+            'zip_plus4': 12,
+            'home_phone': 13,
+            'cell_phone': 14,
+            'business_phone': 15,
+            'other_phone': 16,
+            'fax': 17,
+            'email': 18,
+            'seat_no': 19,
+            'house_leadership': 20,
+            'committees': 21
+        }
 
-            # Get the photo url.
-            html = self.get(leg_url).text
-            doc = lxml.html.fromstring(html)
-            doc.make_links_absolute(leg_url)
-            (photo_url,) = doc.xpath('//img[contains(@src, ".jpg")]/@src')
+        url = (
+            'http://legislature.maine.gov/house/128house.xlsx'
+        )
+        fn, result = self.urlretrieve(url)
 
-            # Add contact information from personal page
-            office_address = re.search(
-                r'<B>Address:  </B>(.+?)\n?</?P>', html, re.IGNORECASE).group(1)
+        wb = xlrd.open_workbook(fn)
+        sh = wb.sheet_by_index(0)
 
-            office_email = doc.xpath(
-                '//a[starts-with(@href, "mailto:")]/text()')
-            business_phone = re.search(
-                r'<B>Business Telephone:  </B>(.+?)</?P>', html, re.IGNORECASE)
-            home_phone = re.search(
-                r'<B>Home Telephone:  </B>(.+?)</?P>', html, re.IGNORECASE)
-            cell_phone = re.search(
-                r'<B>Cell Telephone:  </B>(.+?)</?P>', html, re.IGNORECASE)
+        for rownum in range(1, sh.nrows):
+            # get fields out of mapping
+            d = {}
+            for field, col_num in mapping.items():
+                try:
+                    d[field] = str(sh.cell(rownum, col_num).value).strip()
+                except UnicodeEncodeError:
+                    # str typecasting will raise error for non-ascii characters.
+                    d[field] = sh.cell(rownum, col_num).value.encode('utf-8').strip()
+                except IndexError:
+                    # This col_num doesn't exist in the sheet.
+                    pass
+
+            district_number = d['district']
+            district_name = d['legal_res']
+
+            first_name = d['first_name']
+            middle_name = d['mid_name']
+            last_name = d['last_name']
+            jr_sr = d['jr_sr']
+
+            member_name = " ".join((first_name, middle_name, last_name, jr_sr))
+            member_name = re.sub(r'\s+', ' ', member_name).strip()
+
+            party = d['party']
+            party = _party_map[party]
+
+            # Determine legislator's URL
+            leg_url = 'http://legislature.maine.gov/house/hsebios/'
```

Comment:
Does this pattern _always_ work? If it's iffy, might be worth adding something like this afterwards, to help debug or build in the future:

```
try:
    self.head(leg_url)
except scrapelib.HTTPError:
    self.warning("Could not correctly guess legislator URL for {}".format(member_name))
```

Reply:
Okay, I realised this is a **very** bad idea. Works most of the times, but sometimes the urls are being formed randomly, taking different length combinations from different parts of the name. I've changed my approach, falling back on the way it was being done previously for the upper house, ie, finding the url from the house roster. This seems to work for everyone, other than for Owen D. Casás. This has something to do with the special character, because of which lxml cannot find the element